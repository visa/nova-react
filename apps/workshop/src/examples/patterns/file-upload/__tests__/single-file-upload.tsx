/**
 *              Â© 2025 Visa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 **/
import { act, fireEvent, render, screen, within, cleanup } from '@testing-library/react';
import { axe } from 'jest-axe';
import SingleFileUpload, { MAX_FILE_SIZE, ACCEPTED_FILE_TYPES } from '../single-file-upload.tsx';

import { mockUpload } from '../shared/mock-upload.ts';

jest.mock('../shared/mock-upload', () => ({
  mockUpload: jest.fn(),
}));

describe('SingleFileUpload', () => {
  const mockUploadFn = mockUpload as jest.MockedFunction<typeof mockUpload>;

  afterEach(() => {
    cleanup();
  });

  it('should render correctly', async () => {
    const { container } = render(<SingleFileUpload />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });

  it('should trigger file input when Select file button is clicked', () => {
    const { container } = render(<SingleFileUpload />);

    const fileInput = container.querySelector('input[type="file"]') as HTMLInputElement;
    const selectButton = screen.getByText('Select file');

    const clickSpy = jest.spyOn(fileInput, 'click');

    fireEvent.click(selectButton);

    expect(clickSpy).toHaveBeenCalled();
  });

  it('should process files when files are selected via input', async () => {
    jest.useFakeTimers();
    mockUploadFn.mockResolvedValue();

    const { container } = render(<SingleFileUpload />);

    const fileInput = container.querySelector('input[type="file"]') as HTMLInputElement;
    const file = new File(['test content'], 'test.pdf', { type: 'application/pdf' });

    await act(async () => {
      fireEvent.change(fileInput, { target: { files: [file] } });
    });

    expect(screen.getByText('test.pdf')).toBeInTheDocument();
    expect(mockUploadFn).toHaveBeenCalledWith(
      expect.objectContaining({
        file: expect.objectContaining({
          name: 'test.pdf',
          type: 'application/pdf',
        }),
      }),
      {
        maxFileSize: MAX_FILE_SIZE,
        acceptedFileTypes: ACCEPTED_FILE_TYPES.map(t => t.type),
      }
    );
  });

  it('should retry file upload when retry button is clicked', async () => {
    jest.useFakeTimers();
    mockUploadFn.mockRejectedValueOnce(new Error('Upload failed')).mockResolvedValueOnce();

    const { container } = render(<SingleFileUpload />);

    const fileInput = container.querySelector('input[type="file"]') as HTMLInputElement;
    const file = new File(['test content'], 'test.pdf', { type: 'application/pdf' });

    await act(async () => {
      fireEvent.change(fileInput, { target: { files: [file] } });
    });

    const retryButton = await within(container).findByLabelText(/^Retry upload/);

    fireEvent.click(retryButton);

    await act(async () => {
      jest.runOnlyPendingTimers();
    });

    const fileListItem = within(container).getByRole('listitem');
    expect(within(fileListItem).getByLabelText(/^Success/)).toBeInTheDocument();
  });
});
