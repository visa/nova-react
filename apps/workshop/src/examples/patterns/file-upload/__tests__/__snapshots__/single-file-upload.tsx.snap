// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleFileUpload should render correctly 1`] = `
<div>
  <div
    class="v-flex v-flex-col v-gap-25"
  >
    <div
      class="v-flex v-flex-col v-gap-25"
      style="max-width: 400px;"
    >
      <div
        class="v-flex v-flex-col v-gap-16"
      >
        <div
          class="v-flex v-flex-col v-gap-8"
        >
          <h4
            class="v-typography-subtitle-1"
          >
            Upload file
          </h4>
          <p
            class="v-typography-label-small"
          >
            Choose one file to upload, up to 25 MB each. Accepted file types are .pdf, .png, and .jpg.
          </p>
        </div>
        <input
          class="v-screen-reader"
          hidden=""
          type="file"
        />
        <button
          class="v-button v-button-secondary v-align-self-start"
        >
          Select file
        </button>
      </div>
      <div
        class="v-screen-reader"
        role="status"
      />
    </div>
    <div
      class="v-align-self-end"
      role="alert"
    />
  </div>
</div>
`;
